lib/pbench/server/db/models/tracker.py

```diff
@@ -0,0 +1,299 @@
+import datetime
+import enum
+import os
+
+from sqlalchemy import Column, Integer, String, DateTime, Enum
+
+from pbench.server.db.database import Database
+from sqlalchemy.exc import SQLAlchemyError
+
+
+class DatasetError(Exception):
+    pass
+
+
+class DatasetSqlError(DatasetError):
+    pass
+
+
+class DatasetNotFound(DatasetError):
+    pass
+
+
+class DatasetTransitionError(DatasetError):
+    pass
+
+
+class DatasetTerminalStateViolation(DatasetTransitionError):
+    pass
+
+
+class DatasetBadStateTransition(DatasetTransitionError):
+    pass
+
+
+class States(enum.Enum):
+    """
+    States: Track the progress of a dataset (tarball) through the various
+    stages and operation of the Pbench server.
+    """
+
+    UPLOADING = ("Uploading", True)
+    UPLOADED = ("Uploaded", False)
+    ARCHIVING = ("Archiving", True)
+    ARCHIVED = ("Archived", False)
+    UNPACKING = ("Unpacking", True)
+    UNPACKED = ("Unpacked", False)
+    INDEXING = ("Indexing", True)
+    INDEXED = ("Indexed", False)
+    REINDEXING = ("Re-indexing", True)
```

Comment:
I actually had something more squirrelly in mind, but I wrote a "state lifecycle" pytest case yesterday that read the transition table and made sure we could move a dataset through the entire set. The problem is that `REINDEXING` sets up a loop, from which I'd need bizarre special logic to escape. My original idea had been REINDEXING -> INDEXING -> INDEXED, but the more I think about it, the less I like it.

For one thing, I'd like to trigger indexing by a query for datasets in a starting state; but having `pbench-reindex.py` set the dataset state back to `UNPACKED` would be weird. It might make more sense to do this as a metadata; and `pbench-index.py` would select "all datasets in state `UNPACKED` or with a `reindex` metadata tag. In which case I could just get rid of the `REINDEXING` state entirely. Since re-index is an odd special case, I could even ignore that for the time being rather than adding the metadata mechanism now.

Still thinking about this one...

Reply:
Why does it add a loop of re-indexing transitions to indexed?

Reply:
> Why does it add a loop of re-indexing transitions to indexed?

Because the test case reads the `transitions` table and applies the first state in the value array for the current state:

```
        States.INDEXED: [States.REINDEXING, States.EXPIRING],
        States.REINDEXING: [States.INDEXED],
```

It was kinda funny the first time. I hacked around it by stopping the iteration when I saw a current state I'd already used, but that means the test case is never moving on to `EXPIRING`. I just didn't bother to think through a better solution at the time; but I will before I'm done. Anyway, removing the `REINDEXING` state would make sense, logically, because it _isn't_ really a distinct state in the first place. Pbench doesn't even really "reindex" anyway; we just push some buttons to trigger another `pbench-index.py` as if we'd never done it before.